### PR TITLE
[SPARK-12617][PySpark]Clean up the leak sockets of Py4J

### PR DESCRIPTION
**Pull request by zsxwing**
This patch added Py4jCallbackConnectionCleaner to clean the leak sockets of Py4J every 30 seconds. This is a workaround before Py4J fixes the leak issue https://github.com/bartdag/py4j/issues/187

### CODE REVIEW

**Comment by zsxwing** [comment]
CC @davies

File: python/pyspark/context.py [hunk 1]
```diff
@@ -54,6 +54,63 @@
...
+
+    def _start_timer(self, f):
+        from threading import Timer
+
+        with self._lock:
+            if not self._stopped:
+                self._timer = Timer(30.0, f)
+                self._timer.start()
+
+    def stop(self):
```
Comment (zsxwing) [inline python/pyspark/context.py]: This method is not called because PySpark never stops the gateway server.

**Comment by zsxwing** [comment]
retest this please

**Comment by davies** [comment]
LGTM

**Comment by davies** [comment]
Merged into master and 1.5, 1.6 branch

File: python/pyspark/context.py [hunk 2]
```diff
@@ -54,6 +54,64 @@
...
+
+    def start(self):
+        if self._stopped:
+            return
+
+        def clean_closed_connections():
+            from py4j.java_gateway import quiet_close, quiet_shutdown
+
+            callback_server = self._gateway._callback_server
+            with callback_server.lock:
```
Comment (davies) [inline python/pyspark/context.py]: Traceback (most recent call last):
  File "//anaconda/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "//anaconda/lib/python2.7/threading.py", line 1082, in run
    self.function(*self.args, **self.kwargs)
  File "/Users/davies/work/spark/python/pyspark/context.py", line 78, in clean_closed_connections
    with callback_server.lock:
AttributeError: 'NoneType' object has no attribute 'lock'

Comment (zsxwing) [inline python/pyspark/context.py]: Sorry that I didn't notice that only Streaming uses the callback server. I sent #10621 to fix it.

**Pull request closed without merging**